### ISSUE TITLE
Incorrect error message location

### ISSUE THREAD

**Issue by s3bk** [rust-lang/rust#70082]
I tried this code:
```rust
// the error is reported here
let em = |v: f64| -> f64 { v };

let x: f64 = em(1i16.into());

// but originates here
let y: f64 = 0.01f64 * 1i16.into();
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=75e5444eec728a631e6c26a532acce63))

The error message:
```
error[E0284]: type annotations needed for `f64`
 --> src/main.rs:3:15
  |
3 |     let em = |v: f64| -> f64 { v };
  |               ^ consider giving this closure parameter a type
  |
  = note: cannot resolve `<f64 as std::ops::Mul<_>>::Output == f64`
```

I expected to see this happen: The error is reported on the line for `y`.

Instead, this happened: The error is reported for the closure…

### Meta
Nighty, Beta and Stable affected.

**Comment by bjorn3** [comment]
@rustbot modify labels: +A-diagnostics +C-bug +D-incorrect

(someone other than the issue creator do this?)

**Comment by tirr-c** [comment]
More minimal repro:

```rust
fn main() {
    |_: f64| {};
    let _: f64 = 0.0 * 0.into();
}
```

The closure just needs to be exist anywhere inside `main`. Even this code reports the same error:

```rust
fn main() {
    |_: f64| -> f64 {
        0.0 * 0.into()
    };
}
```

**Comment by contrun** [comment]
This is essential the same issue as https://github.com/rust-lang/rust/issues/69455.
